File: Leetcode/75-Sort-Colors/SortColors.go
func sortColors(nums []int)  {
    if nums == nil || len(nums)<2 {
        return
    }
    cur:=0
    left:=0
    length:=len(nums)
    right:=length-1
    for left<length {
        for left<length && nums[left]==cur{
            left+=1
        }
        for right>=0 && nums[right]!=cur{
            right-=1
        }
        if left<right{
            swap(nums, left, right)
            left+=1
            right-=1
        } else {
            right = length-1
            cur+=1
        }
    }
}

func swap(nums []int, i int, j int) {
    nums[i], nums[j] = nums[j], nums[i]
}

File: Leetcode/75-Sort-Colors/SortColors.cpp
class Solution {
public:
    void sortColors(vector<int>& nums) {
        if (nums.size()<2) {
            return;
        }
        int cur = 0;
        int left = 0;
        int right = nums.size()-1;
        while (left<nums.size()) {
            while (left<nums.size() && nums[left]==cur) {
                left++;
            }
            while (right>=0 && nums[right]!=cur) {
                right--;
            }
            if (left<right) {
                swap(nums, left, right);
                left++;
                right--;
            } else {
                right = nums.size()-1;
                cur++;
            }
        }
    }

private:
    void swap(vector<int>& nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
};
